File: src/main/kotlin/xingkong/loopicserver/picproc/tagImg/match/MatchFloorProperSubnorm.java
package xingkong.loopicserver.picproc.tagImg.match;


import xingkong.loopicserver.picproc.tagImg.TagedFile;

import java.util.List;
import java.util.Random;

public class MatchFloorProperSubnorm extends MatchBase {

    public static final float PROPOR_STEP = 3.0f;

    /**
     * 计算匹配分的函数；使用先验标签匹配法
     *
     * @param tagedFile
     * @param inputTag
     */
    @Override
    public void computeMp(TagedFile tagedFile, List<String> inputTag) {
        int inputTagSize = inputTag.size();
        float tagDis = 1.0f;
        for (int i = tagedFile.getTags().size() - 1; i > 0; i--) {
            String tag = tagedFile.getTags().get(i);
            if (inputTag.contains(tag)) {
                if (i < tagedFile.getTags().size() - 1) {
                    tagDis += PROPOR_STEP;
                }
                tagDis += PROPOR_STEP * (i - inputTagSize);
                break;
            }
        }

        float rand = new Random().nextFloat() * tagDis;
        tagedFile.setMp(rand);   //使用随机数来shuffle
    }
}

File: src/main/kotlin/xingkong/loopicserver/module/match/MatchFloorProperSubnorm.kt
package xingkong.loopicserver.module.match


import xingkong.loopicserver.module.bean.TagedFile
import java.util.*

class MatchFloorProperSubnorm : MatchBase() {

    /**
     * 计算匹配分的函数；使用先验标签匹配法
     *
     * @param tagedFile
     * @param inputTag
     */
    override fun computeMp(tagedFile: TagedFile, inputTag: List<String>) {
        val inputTagSize = inputTag.size
        var tagDis = 1.0f
        for (i in tagedFile.tags.size - 1 downTo 1) {
            val tag = tagedFile.tags[i]
            if (inputTag.contains(tag)) {
                if (i < tagedFile.tags.size - 1) {
                    tagDis += PROPOR_STEP
                }
                tagDis += PROPOR_STEP * (i - inputTagSize)
                break
            }
        }

        val rand = Random().nextDouble() * tagDis
        tagedFile.mp = rand   //使用随机数来shuffle
    }

    companion object {

        val PROPOR_STEP = 3.0f
    }
}
